// ts-to-json/dist/src/NodeParser/TypeReferenceNodeParser.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypeReferenceNodeParser = void 0;
const typescript_1 = __importDefault(require("typescript"));
const UnknownSymbolDeclarations_1 = require("../Error/UnknownSymbolDeclarations");
const UnknownTypeReference_1 = require("../Error/UnknownTypeReference");
const NodeParser_1 = require("../NodeParser");
const AnnotatedType_1 = require("../Type/AnnotatedType");
const ArrayType_1 = require("../Type/ArrayType");
const StringType_1 = require("../Type/StringType");
const UnknownSymbolType_1 = require("../Type/UnknownSymbolType");
const invalidTypes = {
    [typescript_1.default.SyntaxKind.ModuleDeclaration]: true,
    [typescript_1.default.SyntaxKind.VariableDeclaration]: true,
};
class TypeReferenceNodeParser {
    constructor(typeChecker, childNodeParser, config) {
        this.typeChecker = typeChecker;
        this.childNodeParser = childNodeParser;
        this.config = config;
    }
    supportsNode(node) {
        return node.kind === typescript_1.default.SyntaxKind.TypeReference;
    }
    createType(node, context) {
        const typeSymbol = this.typeChecker.getSymbolAtLocation(node.typeName);
        if (!typeSymbol || typeSymbol.name === "unknown") {
            if (this.config.handleUnknownTypes)
                return new UnknownSymbolType_1.UnknownSymbolType(node, typeSymbol, this.config.allowArbitraryDataTypes);
            throw new UnknownTypeReference_1.UnknownTypeReference(node, `are you using global types?`);
        }
        if (typeSymbol.flags & typescript_1.default.SymbolFlags.Alias) {
            const aliasedSymbol = this.typeChecker.getAliasedSymbol(typeSymbol);
            if (aliasedSymbol.name === "unknown") {
                if (this.config.handleUnknownTypes)
                    return new UnknownSymbolType_1.UnknownSymbolType(node, aliasedSymbol, this.config.allowArbitraryDataTypes);
                throw new UnknownTypeReference_1.UnknownTypeReference(node);
            }
            return this.childNodeParser.createType(aliasedSymbol.declarations.filter((n) => !invalidTypes[n.kind])[0], this.createSubContext(node, context));
        }
        else if (typeSymbol.flags & typescript_1.default.SymbolFlags.TypeParameter) {
            return context.getArgument(typeSymbol.name);
        }
        else if (typeSymbol.name === "Array" || typeSymbol.name === "ReadonlyArray") {
            const type = this.createSubContext(node, context).getArguments()[0];
            if (type === undefined) {
                return undefined;
            }
            return new ArrayType_1.ArrayType(type);
        }
        else if (typeSymbol.name === "Date") {
            return new AnnotatedType_1.AnnotatedType(new StringType_1.StringType(), { format: "date-time" }, false);
        }
        else if (typeSymbol.name === "RegExp") {
            return new AnnotatedType_1.AnnotatedType(new StringType_1.StringType(), { format: "regex" }, false);
        }
        else {
            if (typeSymbol.declarations) {
                return this.childNodeParser.createType(typeSymbol.declarations.filter((n) => !invalidTypes[n.kind])[0], this.createSubContext(node, context));
            }
            else {
                if (this.config.handleUnknownTypes)
                    return new UnknownSymbolType_1.UnknownSymbolType(node, typeSymbol, this.config.allowArbitraryDataTypes);
                throw new UnknownSymbolDeclarations_1.UnknownSymbolDeclarations(node);
            }
        }
    }
    createSubContext(node, parentContext) {
        const subContext = new NodeParser_1.Context(node, parentContext);
        if (node.typeArguments && node.typeArguments.length) {
            for (const typeArg of node.typeArguments) {
                const type = this.childNodeParser.createType(typeArg, parentContext);
                subContext.pushArgument(type);
            }
        }
        return subContext;
    }
}
exports.TypeReferenceNodeParser = TypeReferenceNodeParser;
