// ts-to-json/dist/src/NodeParser/UnionNodeParser.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnionNodeParser = void 0;
const typescript_1 = __importDefault(require("typescript"));
const UnionType_1 = require("../Type/UnionType");
const notUndefined_1 = require("../Utils/notUndefined");
const isExcludedProp_1 = require("../Utils/isExcludedProp");
class UnionNodeParser {
    constructor(typeChecker, childNodeParser, config) {
        this.typeChecker = typeChecker;
        this.childNodeParser = childNodeParser;
        this.config = config;
    }
    supportsNode(node) {
        return node.kind === typescript_1.default.SyntaxKind.UnionType;
    }
    createType(node, context) {
        const types = node.types
            .filter((x) => !(0, isExcludedProp_1.isExcludedProp)(x, context, this.config))
            .map((subnode) => {
            return this.childNodeParser.createType(subnode, context);
        })
            .filter(notUndefined_1.notUndefined);
        if (types.length === 1) {
            return types[0];
        }
        else if (types.length === 0) {
            return undefined;
        }
        return new UnionType_1.UnionType(types);
    }
}
exports.UnionNodeParser = UnionNodeParser;
