// ts-to-json/dist/src/Type/AnnotatedType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnnotatedType = void 0;
const BaseType_1 = require("./BaseType");
const nodeKey_1 = require("../Utils/nodeKey");
const deepMerge2_1 = require("../Utils/deepMerge2");
const AliasType_1 = require("./AliasType");
const getAnnotations = (data, annotations = {}) => {
    const annotation = data.annotations;
    if (data.type && (data.type instanceof AnnotatedType || data.type instanceof AliasType_1.AliasType)) {
        const refAnnotations = getAnnotations(data.type, annotation);
        const mergedAnnotations = (0, deepMerge2_1.deepMerge)(refAnnotations, annotation || {});
        return mergedAnnotations;
    }
    return annotation;
};
class AnnotatedType extends BaseType_1.BaseType {
    constructor(type, annotations, nullable) {
        super();
        this.type = type;
        this.annotations = annotations;
        this.nullable = nullable;
    }
    getId() {
        return this.type.getId() + (0, nodeKey_1.hash)([this.isNullable(), this.annotations]);
    }
    getType() {
        return this.type;
    }
    getAnnotations() {
        return getAnnotations(this);
    }
    isNullable() {
        return this.nullable;
    }
}
exports.AnnotatedType = AnnotatedType;
