// ts-to-json/dist/src/Type/FunctionType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FunctionType = exports.FunctionParameter = void 0;
const BaseType_1 = require("./BaseType");
class FunctionParameter {
    constructor(name, type, required) {
        this.name = name;
        this.type = type;
        this.required = required;
    }
    getName() {
        return this.name;
    }
    getType() {
        return this.type;
    }
    isRequired() {
        return this.required;
    }
}
exports.FunctionParameter = FunctionParameter;
class FunctionType extends BaseType_1.BaseType {
    constructor(id, baseTypes, properties, additionalParameters, returnType) {
        super();
        this.id = id;
        this.baseTypes = baseTypes;
        this.properties = properties;
        this.additionalParameters = additionalParameters;
        this.returnType = returnType;
    }
    getId() {
        return this.id;
    }
    getReturnType() {
        return this.returnType;
    }
    getBaseTypes() {
        return this.baseTypes;
    }
    getParameters() {
        return this.properties;
    }
    getAdditionalParameters() {
        return this.additionalParameters;
    }
}
exports.FunctionType = FunctionType;
