// ts-to-json/dist/src/CircularReferenceNodeParser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CircularReferenceNodeParser = void 0;
const ReferenceType_1 = require("./Type/ReferenceType");
const Utils_1 = require("./Utils");
class CircularReferenceNodeParser {
    constructor(childNodeParser, config) {
        this.childNodeParser = childNodeParser;
        this.config = config;
        this.circular = new Map();
    }
    supportsNode(node) {
        if (!(0, Utils_1.shouldParseNode)(node, this.config)) {
            if (!(0, Utils_1.isInForceParseTypes)(node, this.config)) {
                return false;
            }
        }
        return this.childNodeParser.supportsNode(node);
    }
    createType(node, context) {
        const key = (0, Utils_1.extendKey)((0, Utils_1.getKey)(node, context), node, context, this.config);
        context.ignoreLimits = (0, Utils_1.hasLimitOptions)(this.config) && (0, Utils_1.isRecursionToType)(node, context, this.config.type);
        if (this.circular.has(key)) {
            context.ignoreLimits = false;
            return this.circular.get(key);
        }
        const reference = new ReferenceType_1.ReferenceType();
        reference.setId("circularRef-" + key);
        this.circular.set(key, reference);
        const type = this.childNodeParser.createType(node, context, reference);
        context.ignoreLimits = false;
        if (type) {
            reference.setType(type);
        }
        this.circular.delete(key);
        return type;
    }
}
exports.CircularReferenceNodeParser = CircularReferenceNodeParser;
